File: Python 1 Arquivo/Principal.py
import time
from Util import Util

class Ordenacao:
    @staticmethod
    def bolha(lista):
        houve_troca = True
        qtd_comparacoes = 0
        qtd_trocas = 0
        while houve_troca:
            houve_troca = False
            for i in range(len(lista) - 1):
                qtd_comparacoes += 1
                if lista[i] > lista[i + 1]:
                    qtd_trocas += 1
                    houve_troca = True
                    lista[i], lista[i + 1] = lista[i + 1], lista[i]
        return qtd_comparacoes, qtd_trocas

    @staticmethod
    def selecao(lista):
        qtd_comparacoes = 0
        qtd_trocas = 0
        for i in range(len(lista) - 1):
            posMenor = i
            for j in range(i + 1, len(lista)):
                qtd_comparacoes += 1
                if lista[j] < lista[posMenor]:
                    posMenor = j
            if i != posMenor:
                qtd_trocas += 1
                lista[i], lista[posMenor] = lista[posMenor], lista[i]
        return qtd_comparacoes, qtd_trocas

    @staticmethod
    def insercao(lista):
        qtd_comparacoes = 0
        qtd_trocas = 0
        for i in range(1, len(lista)):
            tmp = lista[i]
            j = i - 1
            while j >= 0 and tmp < lista[j]:
                qtd_comparacoes += 1
                lista[j + 1] = lista[j]
                qtd_trocas += 1
                j -= 1
            qtd_comparacoes += 1
            lista[j + 1] = tmp
            qtd_trocas += 1
        return qtd_comparacoes, qtd_trocas

    @staticmethod
    def agitacao(lista):
        ini = 0
        fim = len(lista) - 1
        qtd_comparacoes = 0
        qtd_trocas = 0
        while True:
            houve_troca = False
            for i in range(ini, fim):
                qtd_comparacoes += 1
                if lista[i] > lista[i + 1]:
                    qtd_trocas += 1
                    lista[i], lista[i + 1] = lista[i + 1], lista[i]
                    houve_troca = True
            if not houve_troca:
                break
            fim -= 1
            houve_troca = False
            for i in range(fim, ini, -1):
                qtd_comparacoes += 1
                if lista[i] < lista[i - 1]:
                    qtd_trocas += 1
                    lista[i], lista[i - 1] = lista[i - 1], lista[i]
                    houve_troca = True
            ini += 1
            if not houve_troca:
                break
        return qtd_comparacoes, qtd_trocas

    @staticmethod
    def pente(lista):
        distancia = len(lista)
        houve_troca = True
        qtd_comparacoes = 0
        qtd_trocas = 0
        while houve_troca or distancia > 1:
            distancia = int(distancia / 1.3)
            if distancia < 1:
                distancia = 1
            houve_troca = False
            for i in range(len(lista) - distancia):
                qtd_comparacoes += 1
                if lista[i] > lista[i + distancia]:
                    qtd_trocas += 1
                    lista[i], lista[i + distancia] = lista[i + distancia], lista[i]
                    houve_troca = True
        return qtd_comparacoes, qtd_trocas

    @staticmethod
    def particiona(lista, ini, fim, cont):
        pivo = ini
        while fim > ini:
            while fim > pivo and lista[fim] > lista[pivo]:
                cont[0] += 1
                fim -= 1
            if fim > pivo:
                cont[1] += 1
                lista[pivo], lista[fim] = lista[fim], lista[pivo]
                pivo = fim
            ini += 1
            while ini < pivo and lista[ini] < lista[pivo]:
                cont[0] += 1
                ini += 1
            if ini < pivo:
                cont[1] += 1
                lista[pivo], lista[ini] = lista[ini], lista[pivo]
                pivo = ini
        return pivo

    @staticmethod
    def quickSort(lista):
        cont = [0, 0]
        def _quickSort(lista, ini, fim, cont):
            if ini < fim:
                pivo = Ordenacao.particiona(lista, ini, fim, cont)
                if ini < pivo - 1:
                    _quickSort(lista, ini, pivo - 1, cont)
                if pivo + 1 < fim:
                    _quickSort(lista, pivo + 1, fim, cont)
        _quickSort(lista, 0, len(lista) - 1, cont)
        return cont[0], cont[1]


# ---------------- MAIN ----------------

lista_bolha = []
lista_normal = []
lista_selecao = []
lista_insercao = []
lista_agitacao = []
lista_pente = []
lista_quick = []

tamanho = 1000

Util.popular_lista_aleatoria(lista_bolha, tamanho, 10000, 20000)
lista_normal.extend(lista_bolha)
lista_selecao.extend(lista_bolha)
lista_insercao.extend(lista_bolha)
lista_agitacao.extend(lista_bolha)
lista_pente.extend(lista_bolha)
lista_quick.extend(lista_bolha)

tempoInicio = time.perf_counter()
lista_normal.sort()
tempoFim = time.perf_counter()
print("Tempo da rotina ordenar por sort nativo: ", (tempoFim - tempoInicio), "s")

tempoInicio = time.perf_counter()
qtd_comparacoes, qtd_trocas = Ordenacao.bolha(lista_bolha)
tempoFim = time.perf_counter()
print("Tempo da rotina ordenar por bolha: ", (tempoFim - tempoInicio), "s")
print("Comparacoes:", qtd_comparacoes)
print("Trocas:", qtd_trocas)

tempoInicio = time.perf_counter()
qtd_comparacoes, qtd_trocas = Ordenacao.selecao(lista_selecao)
tempoFim = time.perf_counter()
print("Tempo da rotina ordenar por selecao: ", (tempoFim - tempoInicio), "s")
print("Comparacoes:", qtd_comparacoes)
print("Trocas:", qtd_trocas)

tempoInicio = time.perf_counter()
qtd_comparacoes, qtd_trocas = Ordenacao.insercao(lista_insercao)
tempoFim = time.perf_counter()
print("Tempo da rotina ordenar por insercao: ", (tempoFim - tempoInicio), "s")
print("Comparacoes:", qtd_comparacoes)
print("Trocas:", qtd_trocas)

tempoInicio = time.perf_counter()
qtd_comparacoes, qtd_trocas = Ordenacao.agitacao(lista_agitacao)
tempoFim = time.perf_counter()
print("Tempo da rotina ordenar por agitacao: ", (tempoFim - tempoInicio), "s")
print("Comparacoes:", qtd_comparacoes)
print("Trocas:", qtd_trocas)

tempoInicio = time.perf_counter()
qtd_comparacoes, qtd_trocas = Ordenacao.pente(lista_pente)
tempoFim = time.perf_counter()
print("Tempo da rotina ordenar por pente: ", (tempoFim - tempoInicio), "s")
print("Comparacoes:", qtd_comparacoes)
print("Trocas:", qtd_trocas)

tempoInicio = time.perf_counter()
qtd_comparacoes, qtd_trocas = Ordenacao.quickSort(lista_quick)
tempoFim = time.perf_counter()
print("Tempo da rotina ordenar por quick: ", (tempoFim - tempoInicio), "s")
print("Comparacoes:", qtd_comparacoes)
print("Trocas:", qtd_trocas)

File: Java/Principal.java
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

class Util {
    public static void popularListaAleatoria(List<Integer> lista, int quantidade, int inicio, int fim) {
        Random random = new Random();
        for (int i = 0; i < quantidade; i++) {
            lista.add(random.nextInt(fim - inicio + 1) + inicio);
        }
    }

    public static void exibirLista(List<Integer> lista, String frase) {
        System.out.println(frase);
        for (int item : lista) {
            System.out.println(item);
        }
    }
}

class Ordenacao {
    // Contadores do Quick (colocados dentro da classe)
    public static int qtdQuickComparacoes = 0;
    public static int qtdQuickTrocas = 0;

    public static int[] bolha(List<Integer> lista) {
        boolean houveTroca = true;
        int qtdComparacoes = 0, qtdTrocas = 0;

        while (houveTroca) {
            houveTroca = false;
            for (int i = 0; i < lista.size() - 1; i++) {
                qtdComparacoes++;
                if (lista.get(i) > lista.get(i + 1)) {
                    int tmp = lista.get(i);
                    lista.set(i, lista.get(i + 1));
                    lista.set(i + 1, tmp);
                    houveTroca = true;
                    qtdTrocas++;
                }
            }
        }
        return new int[]{qtdComparacoes, qtdTrocas};
    }

    public static int[] selecao(List<Integer> lista) {
        int qtdComparacoes = 0, qtdTrocas = 0;

        for (int i = 0; i < lista.size() - 1; i++) {
            int posMenor = i;
            for (int j = i + 1; j < lista.size(); j++) {
                qtdComparacoes++;
                if (lista.get(j) < lista.get(posMenor)) {
                    posMenor = j;
                }
            }
            if (i != posMenor) {
                int tmp = lista.get(i);
                lista.set(i, lista.get(posMenor));
                lista.set(posMenor, tmp);
                qtdTrocas++;
            }
        }
        return new int[]{qtdComparacoes, qtdTrocas};
    }

    public static int[] insercao(List<Integer> lista) {
        int qtdComparacoes = 0, qtdTrocas = 0;

        for (int i = 1; i < lista.size(); i++) {
            int tmp = lista.get(i);
            int j = i - 1;

            while (j >= 0) {
                qtdComparacoes++;
                if (tmp < lista.get(j)) {
                    lista.set(j + 1, lista.get(j));
                    qtdTrocas++;
                    j--;
                } else {
                    break;
                }
            }
            lista.set(j + 1, tmp);
            qtdTrocas++;
        }
        return new int[]{qtdComparacoes, qtdTrocas};
    }

    public static int particiona(List<Integer> lista, int ini, int fim) {
        int i = ini;
        int j = fim;
        int pivo = i;

        while (j > i) {
            while (j > pivo) {
                qtdQuickComparacoes++;
                if (lista.get(j) > lista.get(pivo)) {
                    j--;
                } else {
                    break;
                }
            }

            if (j > pivo) {
                int tmp = lista.get(pivo);
                lista.set(pivo, lista.get(j));
                lista.set(j, tmp);
                qtdQuickTrocas++;
                pivo = j;
            }

            i++;
            while (i < pivo) {
                qtdQuickComparacoes++;
                if (lista.get(i) < lista.get(pivo)) {
                    i++;
                } else {
                    break;
                }
            }

            if (i < pivo) {
                int tmp = lista.get(pivo);
                lista.set(pivo, lista.get(i));
                lista.set(i, tmp);
                qtdQuickTrocas++;
                pivo = i;
            }
        }

        return pivo;
    }

    public static void quickSort(List<Integer> lista, int ini, int fim) {
        if (ini >= fim) return;
        int pivo = particiona(lista, ini, fim); 
        if (ini < pivo - 1) quickSort(lista, ini, pivo - 1);
        if (pivo + 1 < fim) quickSort(lista, pivo + 1, fim);
    }

    public static int[] quick(List<Integer> lista) {
        qtdQuickComparacoes = 0;
        qtdQuickTrocas = 0;
        if (lista.size() > 0) {
            quickSort(lista, 0, lista.size() - 1);
        }
        return new int[]{qtdQuickComparacoes, qtdQuickTrocas};
    }
}

public class Principal {
    public static void main(String[] args) {
        List<Integer> listaBolha = new ArrayList<>();
        List<Integer> listaSelecao = new ArrayList<>();
        List<Integer> listaInsercao = new ArrayList<>();
        List<Integer> listaQuick = new ArrayList<>();

        int tamanho = 1000;
        Util.popularListaAleatoria(listaBolha, tamanho, 10000, 20000);
        listaSelecao.addAll(listaBolha);
        listaInsercao.addAll(listaBolha);
        listaQuick.addAll(listaBolha);

        // Bolha
        long inicio = System.nanoTime();
        int[] resBolha = Ordenacao.bolha(listaBolha);
        long fim = System.nanoTime();
        System.out.println("Tempo bolha: " + (fim - inicio) / 1e6 + " ms");
        System.out.println("Comparações: " + resBolha[0] + ", Trocas: " + resBolha[1]);

        // Seleção
        inicio = System.nanoTime();
        int[] resSelecao = Ordenacao.selecao(listaSelecao);
        fim = System.nanoTime();
        System.out.println("Tempo selecao: " + (fim - inicio) / 1e6 + " ms");
        System.out.println("Comparações: " + resSelecao[0] + ", Trocas: " + resSelecao[1]);

        // Inserção
        inicio = System.nanoTime();
        int[] resInsercao = Ordenacao.insercao(listaInsercao);
        fim = System.nanoTime();
        System.out.println("Tempo insercao: " + (fim - inicio) / 1e6 + " ms");
        System.out.println("Comparações: " + resInsercao[0] + ", Trocas: " + resInsercao[1]);

        // Quick
        inicio = System.nanoTime();
        int[] resQuick = Ordenacao.quick(listaQuick); 
        fim = System.nanoTime();
        System.out.println();
        System.out.println("Tempo Quick: " + (fim - inicio) / 1e6 + " ms");
        System.out.println("Comparações: " + resQuick[0] + ", Trocas: " + resQuick[1]);
    }
}
